daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/VariableMap1.scala

```diff
@@ -431,59 +416,57 @@ class VariableMap private (vTable: Map[GlobalQName, ArrayBuffer[VariableInstance
-      variable.state match {
-        case VariableSet => {
-          referringContext.SDE(
-            "Cannot set variable %s twice. State was: %s. Existing value: %s",
-            variable.rd.globalQName,
-            VariableSet,
-            variable.value,
-          )
-        }
+    val variableInstances = vTable(vrd.vmapIndex)
+    val variable = variableInstances.head
+    variable.state match {
+      case VariableSet => {
+        referringContext.SDE(
+          "Cannot set variable %s twice. State was: %s. Existing value: %s",
+          variable.rd.globalQName,
+          VariableSet,
+          variable.value,
+        )
+      }
 
-        case VariableRead => {
+      case VariableRead => {
+
+        /**
+         * TODO: This should be an SDE, but due to a bug (DAFFODIL-1443) in
+         * the way we evaluate escapeSchemes it could lead us to setting the
+         * variable read too early */
+        pstate.SDW(
+          WarnID.VariableSet,
+          "Cannot set variable %s after reading the default value. State was: %s. Existing value: %s",
+          variable.rd.globalQName,
+          VariableSet,
+          variable.value,
+        )
+        variable.setValue(newValue)
+        variable.setState(VariableSet)
+      }
 
+      case _ => {
+        vrd.direction match {
           /**
-           * TODO: This should be an SDE, but due to a bug (DAFFODIL-1443) in
-           * the way we evaluate escapeSchemes it could lead us to setting the
-           * variable read too early */
-          pstate.SDW(
-            WarnID.VariableSet,
-            "Cannot set variable %s after reading the default value. State was: %s. Existing value: %s",
-            variable.rd.globalQName,
-            VariableSet,
-            variable.value,
-          )
-          variable.setValue(newValue)
-          variable.setState(VariableSet)
-        }
-
-        case _ => {
-          vrd.direction match {
-            /**
-             * Due to potential race conditions regarding the setting of
-             * variables via setVariable and default values in combination with
-             * suspensions during unparsing, we only allow the use of either
-             * setVariable statements or a default value when unparsing a
-             * variable.
-             */
-            case VariableDirection.UnparseOnly | VariableDirection.Both
-                if (vrd.maybeDefaultValueExpr.isDefined && variableInstances.get.size > 1) => {
-              // Variable has an unparse direction, a default value, and a
-              // newVariableInstance
-              pstate.SDE(
-                "Variable %s has an unparse direction and a default value, setting the variable may cause race conditions when combined with a forward referencing expression.",
-                varQName,
-              )
-            }
-            case _ => // Do nothing
+           * Due to potential race conditions regarding the setting of
```

Comment:
This needs a bit more explanation. What exactly are you saying can cause a race here? It's all state for one thread. So how can there be a race? There has to be something non-deterministic going on. 

If two suspensions depend on one variable, one is evaluated and gets the value before the set variable actually executes... so the default value has been consumed before the setVariable runs. When you setVariable it will fail. If these can happen in the opposite order, then yes that's a race, and that's due to non-deterministic behavior somewhere in the suspension scheduling.

I realize suspensions get evaluated in a humanly non-predictable order, but it should be deterministic. 

If not we should make it deterministic so that any given scenario is repeatable, and may be incorrect per the setVariable happening after the read of the default has already occurred, but this should happen every time or never, not sporadically. 

This change set is not about this problem however, so we can park this as a separate bug.

Do we check for this and at least warn about it at compile time? To me if you use newVariableInstance, the variable declaration should not have a default value, I think.

Reply:
I didn't actually change this. This is just an indention change. I'm not sure what the comment is suggesting is the problem. Looks like Josh wrote this, @jadams-tresys, do you remember what this is about?

Reply:
I remember working on this for a while and it needs a lengthy test case in order to hit this error message:

https://github.com/apache/daffodil/blob/716e5336621d6ec645eb4a3d3b47b12680a914ff/daffodil-test/src/test/resources/org/apache/daffodil/section07/variables/variables.tdml#L260

I cannot remember if we get indeterminate behavior from this test case if we remove the error check here and I'm struggling to even understand how my test is set up as it has several layers of nesting.  Unfortunately I don't have time to dig into this, but there might be some useful comments in the pull request where this got merged:

https://github.com/apache/daffodil/pull/473

Reply:
That PR has a lot of context for this. 

Clearly not this change set's problem. 

It is quite possible there are bugs in our variables implementation. We don't have enough here to suggest creating a new JIRA ticket about any of this. 

I'm marking this thread "resolved".